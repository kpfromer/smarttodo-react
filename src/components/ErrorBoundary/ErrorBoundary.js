import React, { Component } from 'react';

import Button from '@material-ui/core/Button';
import Raven from 'raven-js';

import styles from './ErrorBoundary.module.css';

export default class ErrorBoundary extends Component {

  state = {
    error: null
  };

  componentDidCatch(error, errorInfo) {
    this.setState({ error });
    if (process.env.NODE_ENV === 'production') {
      Raven.captureException(error, { extra: errorInfo });
    }
  }

  render() {
    if (this.state.error) {
      return (
        <div className={styles.container}>
          <div
            className={styles.content}
          >
            <p>We're sorry â€” something's gone wrong.</p>
            <p>Our team has been notified, but click below to fill out a report.</p>
            <p>Please try refreshing the page.</p>
          </div>
          <Button variant="contained" onClick={() => Raven.lastEventId() && Raven.showReportDialog()}>
            Send Crash Report
          </Button>
        </div>
      );
    }
    return this.props.children;
  }
}
