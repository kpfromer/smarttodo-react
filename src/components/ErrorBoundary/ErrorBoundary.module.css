.container {
  display: flex;
  flex-direction: column;
  align-items: center;
  height: 100%;
}

.content {
  margin: auto;
  text-align: center;
}