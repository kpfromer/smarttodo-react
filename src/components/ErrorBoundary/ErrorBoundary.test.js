import React, { Component } from "react";
import ErrorBoundary from "./ErrorBoundary";
import { SetupComponent } from "react-component-setup";
import raven from 'raven-js';
import Button from '@material-ui/core/Button';

jest.mock('raven-js', () => ({
  captureException: jest.fn(),
  lastEventId: jest.fn(),
  showReportDialog: jest.fn()
}));

const { mount: setup } = SetupComponent({
  component: ErrorBoundary,
  defaultProps: {
    children: (
      <div>
        <h1>Children!</h1>
      </div>
    )
  }
});

describe('ErrorBoundary', () => {
  const originalEnv = { ...process.env };

  afterEach(() => {
    process.env = originalEnv;
    jest.clearAllMocks();
  });

  it('renders without crashing', () => {
    setup();
  });

  describe('no errors', () => {
    it('renders children', () => {
      expect(setup().wrapper).toMatchSnapshot();
    });
  });

  describe('if child error', () => { // TODO: try to refactor code so error is not logged in jest
    class ThrowErrorChild extends Component {
      render() {
        throw new Error('An error!');
      }
    }

    it('reports error to sentry.io on production', () => {
      process.env.NODE_ENV = 'production';
      setup({
        children: (
          <ThrowErrorChild />
        )
      });
      expect(raven.captureException.mock.calls[0]).toMatchSnapshot();
    });
    it('renders error message and crash report button', () => {
      const { wrapper } = setup({
        children: (
          <ThrowErrorChild />
        )
      });
      expect(wrapper).toMatchSnapshot();
    });
    describe('when user clicks on crash report button', () => {
      it('calls sentry.io showReportDialog', () => {
        raven.lastEventId.mockReturnValue('asdfasd');
        const { wrapper } = setup({
          children: (
            <ThrowErrorChild />
          )
        });

        wrapper.find(Button).simulate('click');
        expect(raven.showReportDialog).toHaveBeenCalled();
      });
    });
  });
});