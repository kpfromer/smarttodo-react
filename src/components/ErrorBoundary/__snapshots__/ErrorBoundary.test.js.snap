// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary if child error renders error message and crash report button 1`] = `
<div
  className="container"
>
  <div
    className="content"
  >
    <p>
      We're sorry â€” something's gone wrong.
    </p>
    <p>
      Our team has been notified, but click below to fill out a report.
    </p>
    <p>
      Please try refreshing the page.
    </p>
  </div>
  <button
    className="jss27 jss1 jss12 jss15"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="jss2"
    >
      Send Crash Report
    </span>
    <span
      className="jss30"
    />
  </button>
</div>
`;

exports[`ErrorBoundary if child error reports error to sentry.io on production 1`] = `
Array [
  [Error: An error!],
  Object {
    "extra": Object {
      "componentStack": "
    in ThrowErrorChild (at ErrorBoundary.test.js:53)
    in ErrorBoundary (created by WrapperComponent)
    in WrapperComponent",
    },
  },
]
`;

exports[`ErrorBoundary no errors renders children 1`] = `
<div>
  <h1>
    Children!
  </h1>
</div>
`;
